### PR TITLE
Implement some more WebGL functions

### PR DESCRIPTION
**Pull request by dzbarsky**
r? @ecoal95

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8761)

### CODE REVIEW

**Comment by ecoal95** [comment]
Great! Looks good to me.

This supersedes #8743 right? I guess that one can be closed?

Same thing, waiting for gleam update.

-S-awaiting-review +S-blocked-on-external 

---

Reviewed 6 of 6 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 5 of 5 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8761#-:-K4RmUb7X9NGz3o0B6BG:1295190825)*

File: components/canvas/webgl_paint_task.rs [hunk 1]
```diff
@@ -110,6 +114,12 @@ impl WebGLPaintTask {
                 gl::enable_vertex_attrib_array(attrib_id),
             CanvasWebGLMsg::GetAttribLocation(program_id, name, chan) =>
                 self.attrib_location(program_id, name, chan),
+            CanvasWebGLMsg::GetBufferParameter(target, param_id,chan) =>
```
Comment (frewsxcv) [inline components/canvas/webgl_paint_task.rs]: Missing a space after this comma (as per `mach test-tid`y)

**Comment by nox** [comment]
@ecoal95 It doesn't seem like the same things are implemented in the two PRs.

**Comment by ecoal95** [comment]
@nox I meant this is on top of the other (the other one has just the first tree commits).

**Comment by frewsxcv** [comment]
Lockfiles need to be updated

https://travis-ci.org/servo/servo/jobs/94347070

**Comment by ecoal95** [comment]
@frewsxcv This is pending a gleam update (https://github.com/servo/gleam/pull/53), that's why it does not compile. Hopefully that PR can be reviewed soon.

**Comment by dzbarsky** [comment]
@ecoal95 want to take a look at the vertexAttrib stuff?

**Comment by ecoal95** [comment]
Reviewed 1 of 1 files at r6.
Review status: 2 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/script/dom/webglrenderingcontext.rs, line 903 \[r7\]](https://reviewable.io:443/reviews/servo/servo/8761#-K4tCMAr7uqDcWe6dl8N:-K4tCMAr7uqDcWe6dl8O:1974266710) ([raw file](https://github.com/servo/servo/blob/42621062159a96add7757675c05b645e8cc10faa/components/script/dom/webglrenderingcontext.rs#L903)):*
This logic seems fine to me, but we don't make any length validation, and thus it's unsafe. If a `Float32Array` with less elements than expected is passed here, we access memory we don't own.

I've taken a look to what other browsers do, and Gecko generates an `INVALID_OPERATION` error, while Chromium generates `INVALID_VALUE`.

I'm fine generating any of those two errors, but validating that is necessary .

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8761)*

**Comment by dzbarsky** [comment]
Review status: 2 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/script/dom/webglrenderingcontext.rs, line 903 \[r7\]](https://reviewable.io:443/reviews/servo/servo/8761#-K4tCMAr7uqDcWe6dl8N:-K5--RsK6l4bW98-4RzN:4229029) ([raw file](https://github.com/servo/servo/blob/42621062159a96add7757675c05b645e8cc10faa/components/script/dom/webglrenderingcontext.rs#L903)):*
So we also need to fix http://mxr.mozilla.org/servo/source/components/script/dom/webglrenderingcontext.rs#852

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8761)*

**Comment by dzbarsky** [comment]
We need to publish a new https://crates.io/crates/gleam

**Comment by jdm** [comment]
Published.

**Comment by dzbarsky** [comment]
Hmm, still can't upgrade gleam:

➜  servo git:(gl_functions) ✗ ./mach cargo-update -p gleam -v
components/servo
    Updating registry `https://github.com/rust-lang/crates.io-index`
      Adding libc v0.1.12
ports/cef
    Updating registry `https://github.com/rust-lang/crates.io-index`
      Adding cocoa v0.1.5
      Adding libc v0.1.12
ports/gonk
    Updating registry `https://github.com/rust-lang/crates.io-index`
      Adding libc v0.2.2

And no changes are made to the Cargo.lock files

**Comment by jdm** [comment]
You probably need `--precise 0.2`.

**Comment by dzbarsky** [comment]
➜  servo git:(gl_functions) ✗ ./mach update-cargo -p gleam --precise 0.2  
components/servo
    Updating registry `https://github.com/rust-lang/crates.io-index`
no matching package named `gleam` found (required by `servo`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: ^0.1
ports/cef
    Updating registry `https://github.com/rust-lang/crates.io-index`
no matching package named `gleam` found (required by `embedding`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: ^0.1
ports/gonk
    Updating registry `https://github.com/rust-lang/crates.io-index`
no matching package named `gleam` found (required by `b2s`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: ^0.1

Same thing for 0.2.0

**Comment by jdm** [comment]
Ah, you'll need to change http://mxr.mozilla.org/servo/search?string=gleam+%3D&find=&findi=&filter=^[^\0]*%24&hitlimit=&tree=servo too.

**Comment by dzbarsky** [comment]
@ecoal95 looks like the deps landed so this can go in now, can you r+?

**Comment by ecoal95** [comment]
Sure!

@bors-servo: r+

-S-awaiting-review +S-awaiting-merge 

---

Reviewed 4 of 4 files at r7, 13 of 13 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8761#-:-K5MdPt_UJuuRqeq85Gf:209618176)*

**Comment by ecoal95** [comment]
@dzbarsky Seems like a few more packages need an update:

```
./components/servo/Cargo.lock:1: duplicate versions for package "gleam"
	found dependency on version 0.1.20
	but highest version is 0.2.0
	try upgrading with ./mach cargo-update -p gleam:0.1.20
	The following packages depend on version 0.1.20:
		servo
		cgl
		io-surface
		layers
		offscreen_gl_context
		servo-skia
./ports/cef/Cargo.lock:1: duplicate versions for package "gleam"
	found dependency on version 0.1.20
	but highest version is 0.2.0
	try upgrading with ./mach cargo-update -p gleam:0.1.20
	The following packages depend on version 0.1.20:
		cgl
		io-surface
		layers
		offscreen_gl_context
		servo
		servo-skia
./ports/gonk/Cargo.lock:1: duplicate versions for package "gleam"
	found dependency on version 0.1.20
	but highest version is 0.2.0
	try upgrading with ./mach cargo-update -p gleam:0.1.20
	The following packages depend on version 0.1.20:
		cgl
		io-surface
		layers
		offscreen_gl_context
		servo
		servo-skia
```

**Comment by ecoal95** [comment]
I'm trying to get landed the pending updates for this, so this can land ASAP.

**Comment by ecoal95** [comment]
They seem to have landed \o/. Thanks @frewsxcv!

I made https://github.com/dzbarsky/servo/pull/1 with the changes produced by editing the `components/servo/Cargo.toml` to update gleam and running:

```
./mach cargo-update -p cgl
./mach cargo-update -p io-surface
./mach cargo-update -p layers
./mach cargo-update -p offscreen_gl_context
./mach cargo-update -p servo-skia
```

Which makes tidy pass again.

**Comment by frewsxcv** [comment]
As per the tidy error on travis, looks like gleam needs to be updated

**Comment by ecoal95** [comment]
@dzbarsky Yes, you need to update `components/servo/Cargo.toml`, which has still `gleam = "0.1"`.

**Pull request merged**